src/sage/schemes/hyperelliptic_curves/hyperelliptic_generic.py

```diff
@@ -248,29 +248,260 @@ def is_smooth(self):
+
+        There are no rational points with `x`-coordinate 3::
+
+            sage: H.is_x_coord(3)
+            False
+
+        The function also handles the case when `h(x)` is not zero::
+
+            sage: R.<x> = PolynomialRing(QQ)
+            sage: f = x^5 + x^3 + 1
+            sage: h = x + 1
+            sage: H = HyperellipticCurve(f, h)
+            sage: H.is_x_coord(1)
+            True
+
+        We can perform these operations over finite fields too::
+
+            sage: # needs sage.rings.finite_rings
+            sage: R.<x> = PolynomialRing(GF(163))
+            sage: f = x^7 + x + 1
+            sage: H = HyperellipticCurve(f)
+            sage: H.is_x_coord(13)
+            True
+
+        Including the case of characteristic two::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F.<z4> = GF(2^4)
+            sage: R.<x> = PolynomialRing(F)
+            sage: f = x^7 + x^3 + 1
+            sage: h = x + 1
+            sage: H = HyperellipticCurve(f, h)
+            sage: H.is_x_coord(z4^3 + z4^2 + z4)
+            True
+
+        AUTHORS:
+
+        - Giacomo Pope (2024): adapted from :meth:`lift_x`
+
+        TESTS::
+
+            The `x`-coordinate must be defined over the base field of the curve::
+
+            sage: p = 11
+            sage: F = GF(11)
+            sage: F_ext = GF(11^2)
+            sage: R.<x> = PolynomialRing(F)
+            sage: f = x^7 + x^3 + 1
+            sage: H = HyperellipticCurve(f)
+            sage: H.is_x_coord(F_ext.gen())
+            Traceback (most recent call last):
+            ...
+            TypeError: x must be coercible into the base ring of the curve
+
+        """
+        f, h = self.hyperelliptic_polynomials()
+        K = self.base_ring()
+        try:
+            x = K(x)
+        except (ValueError, TypeError):
+            raise TypeError('x must be coercible into the base ring of the curve')
+
+        # When h is zero then x is a valid coordinate if y2 is square
+        if not h:
+            y2 = f(x)
+            return y2.is_square()
+        # Generic case for h != 0
+        a = f(x)
+        b = h(x)
+        # Special case for char 2
+        if K.characteristic() == 2:
+            R = f.parent()  # Polynomial ring K[x]
+            F = R([-a, b, 1])
+            return bool(F.roots())
+        # Otherwise x is a point on the curve if the determinant is a square
+        D = b*b + 4*a
+        return D.is_square()
+
     def lift_x(self, x, all=False):
-        f, h = self._hyperelliptic_polynomials
-        x += self.base_ring()(0)
-        one = x.parent()(1)
-        if h.is_zero():
+        """
+        Return one or all points with given `x`-coordinate.
+
+        This method is deterministic: It returns the same data each time when called again with the same.
+
+        INPUT:
+
+        - ``x`` -- an element of the base ring of the curve
+
+        - ``all`` (bool, default False) -- if True, return a (possibly
```

Comment:
I think booleans `True` and `False` are commonly wrapped in `` to style them as code.